### PR TITLE
Add tests for ways in which DOMException is unusual

### PR DESCRIPTION
**Pull request by domenic**
See https://github.com/heycam/webidl/issues/55.

DO NOT MERGE; this is in fact meant to demonstrate that nobody implements the current spec, which as discussed in that Web IDL issue I kind of messed up. So, likely we will fix the spec, then switch these tests to be -historical.html tests that test the opposite of their current values.

### CODE REVIEW

**Comment by domenic** [comment]
I've switched these from testing the current-as-of-now spec to testing https://github.com/heycam/webidl/pull/378, which is much closer to what's implemented. The results are now:

- Firefox passes all tests
- Chrome misses the stack property test (which is a "should" in the proposed spec)
- Safari misses the stack property test, and seems to have made message/name/code own properties, like the current-as-of-now spec. But they did not implement any of the other non-interface-like behaviors.
- Edge fails all the tests because they have no DOMException constructor.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/248010580)

# Firefox (nightly) #
Testing web-platform-tests at revision 5b68a1d7318424ca6cdd73a0a3a9842fc97de1ef
Using browser at version BuildID 20170612100241; SourceStamp 27cad9749cddf68e11fdd4e5d73dad84a8f8cf23
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>2 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6361/WebIDL/ecmascript-binding/es-exceptions/DOMException-constructor.html">/WebIDL/ecmascript-binding/es-exceptions/DOMException-constructor.html</a></summary>


|                            Subtest                             | Results | Messages |
|----------------------------------------------------------------|---------|----------|
|                                                                | OK      |          |
| `new DOMException()`                                           | PASS    |          |
| `new DOMException(): inherited-ness`                           | PASS    |          |
| `new DOMException(null)`                                       | PASS    |          |
| `new DOMException(undefined)`                                  | PASS    |          |
| `new DOMException(undefined): inherited-ness`                  | PASS    |          |
| `new DOMException("foo")`                                      | PASS    |          |
| `new DOMException("foo"): inherited-ness`                      | PASS    |          |
| `new DOMException("bar", undefined)`                           | PASS    |          |
| `new DOMException("bar", "NotSupportedError")`                 | PASS    |          |
| `new DOMException("bar", "NotSupportedError"): inherited-ness` | PASS    |          |
| `new DOMException("bar", "foo")`                               | PASS    |          |
| `new DOMexception("msg", "IndexSizeError")`                    | PASS    |          |
| `new DOMexception("msg", "HierarchyRequestError")`             | PASS    |          |
| `new DOMexception("msg", "WrongDocumentError")`                | PASS    |          |
| `new DOMexception("msg", "InvalidCharacterError")`             | PASS    |          |
| `new DOMexception("msg", "NoModificationAllowedError")`        | PASS    |          |
| `new DOMexception("msg", "NotFoundError")`                     | PASS    |          |
| `new DOMexception("msg", "NotSupportedError")`                 | PASS    |          |
| `new DOMexception("msg", "InUseAttributeError")`               | PASS    |          |
| `new DOMexception("msg", "InvalidStateError")`                 | PASS    |          |
| `new DOMexception("msg", "SyntaxError")`                       | PASS    |          |
| `new DOMexception("msg", "InvalidModificationError")`          | PASS    |          |
| `new DOMexception("msg", "NamespaceError")`                    | PASS    |          |
| `new DOMexception("msg", "InvalidAccessError")`                | PASS    |          |
| `new DOMexception("msg", "SecurityError")`                     | PASS    |          |
| `new DOMexception("msg", "NetworkError")`                      | PASS    |          |
| `new DOMexception("msg", "AbortError")`                        | PASS    |          |
| `new DOMexception("msg", "URLMismatchError")`                  | PASS    |          |
| `new DOMexception("msg", "QuotaExceededError")`                | PASS    |          |
| `new DOMexception("msg", "TimeoutError")`                      | PASS    |          |
| `new DOMexception("msg", "InvalidNodeTypeError")`              | PASS    |          |
| `new DOMexception("msg", "DataCloneError")`                    | PASS    |          |
| `new DOMException("bar", "UnknownError")`                      | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6361/WebIDL/ecmascript-binding/es-exceptions/DOMException-custom-bindings.html">/WebIDL/ecmascript-binding/es-exceptions/DOMException-custom-bindings.html</a></summary>


|                                           Subtest                                            | Results | Messages |
|----------------------------------------------------------------------------------------------|---------|----------|
|                                                                                              | OK      |          |
| `Cannot construct without new`                                                               | PASS    |          |
| `inherits from Error: prototype-side`                                                        | PASS    |          |
| `does not inherit from Error: class-side`                                                    | PASS    |          |
| `message property descriptor`                                                                | PASS    |          |
| `name property descriptor`                                                                   | PASS    |          |
| `code property descriptor`                                                                   | PASS    |          |
| `code property is not affected by shadowing the name property`                               | PASS    |          |
| `Object.prototype.toString behavior is like other interfaces`                                | PASS    |          |
| `toString() is defined like for other interfaces, and behaves correctly`                     | PASS    |          |
| `If the implementation has a stack property on normal errors, it also does on DOMExceptions` | PASS    |          |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/248010581)

# Sauce (safari) #
Testing web-platform-tests at revision 5b68a1d7318424ca6cdd73a0a3a9842fc97de1ef
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>2 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6361/WebIDL/ecmascript-binding/es-exceptions/DOMException-constructor.html">/WebIDL/ecmascript-binding/es-exceptions/DOMException-constructor.html</a></summary>


|                            Subtest                             | Results |                                           Messages                                          |
|----------------------------------------------------------------|---------|---------------------------------------------------------------------------------------------|
|                                                                | OK      |                                                                                             |
| `new DOMException()`                                           | FAIL    | `function is not a constructor (evaluating 'new DOMException()')`                           |
| `new DOMException(): inherited-ness`                           | FAIL    | `function is not a constructor (evaluating 'new DOMException()')`                           |
| `new DOMException(null)`                                       | FAIL    | `function is not a constructor (evaluating 'new DOMException(null)')`                       |
| `new DOMException(undefined)`                                  | FAIL    | `function is not a constructor (evaluating 'new DOMException(undefined)')`                  |
| `new DOMException(undefined): inherited-ness`                  | FAIL    | `function is not a constructor (evaluating 'new DOMException(undefined)')`                  |
| `new DOMException("foo")`                                      | FAIL    | `function is not a constructor (evaluating 'new DOMException("foo")')`                      |
| `new DOMException("foo"): inherited-ness`                      | FAIL    | `function is not a constructor (evaluating 'new DOMException("foo")')`                      |
| `new DOMException("bar", undefined)`                           | FAIL    | `function is not a constructor (evaluating 'new DOMException("bar", undefined)')`           |
| `new DOMException("bar", "NotSupportedError")`                 | FAIL    | `function is not a constructor (evaluating 'new DOMException("bar", "NotSupportedError")')` |
| `new DOMException("bar", "NotSupportedError"): inherited-ness` | FAIL    | `function is not a constructor (evaluating 'new DOMException("bar", "NotSupportedError")')` |
| `new DOMException("bar", "foo")`                               | FAIL    | `function is not a constructor (evaluating 'new DOMException("bar", "foo")')`               |
| `new DOMexception("msg", "IndexSizeError")`                    | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "HierarchyRequestError")`             | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "WrongDocumentError")`                | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "InvalidCharacterError")`             | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "NoModificationAllowedError")`        | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "NotFoundError")`                     | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "NotSupportedError")`                 | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "InUseAttributeError")`               | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "InvalidStateError")`                 | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "SyntaxError")`                       | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "InvalidModificationError")`          | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "NamespaceError")`                    | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "InvalidAccessError")`                | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "SecurityError")`                     | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "NetworkError")`                      | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "AbortError")`                        | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "URLMismatchError")`                  | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "QuotaExceededError")`                | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "TimeoutError")`                      | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "InvalidNodeTypeError")`              | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMexception("msg", "DataCloneError")`                    | FAIL    | `function is not a constructor (evaluating 'new DOMException("msg", test_case.name)')`      |
| `new DOMException("bar", "UnknownError")`                      | FAIL    | `function is not a constructor (evaluating 'new DOMException("bar", "UnknownError")')`      |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6361/WebIDL/ecmascript-binding/es-exceptions/DOMException-custom-bindings.html">/WebIDL/ecmascript-binding/es-exceptions/DOMException-custom-bindings.html</a></summary>


|                                           Subtest                                            | Results |                                               Messages                                              |
|----------------------------------------------------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------|
|                                                                                              | OK      |                                                                                                     |
| `Cannot construct without new`                                                               | PASS    |                                                                                                     |
| `inherits from Error: prototype-side`                                                        | PASS    |                                                                                                     |
| `does not inherit from Error: class-side`                                                    | PASS    |                                                                                                     |
| `message property descriptor`                                                                | FAIL    | `function is not a constructor (evaluating 'new DOMException("message", "name")')`                  |
| `name property descriptor`                                                                   | FAIL    | `function is not a constructor (evaluating 'new DOMException("message", "name")')`                  |
| `code property descriptor`                                                                   | FAIL    | `function is not a constructor (evaluating 'new DOMException("message", "name")')`                  |
| `code property is not affected by shadowing the name property`                               | FAIL    | `function is not a constructor (evaluating 'new DOMException("message", "InvalidCharacterError")')` |
| `Object.prototype.toString behavior is like other interfaces`                                | FAIL    | `function is not a constructor (evaluating 'new DOMException("message", "name")')`                  |
| `toString() is defined like for other interfaces, and behaves correctly`                     | FAIL    | `function is not a constructor (evaluating 'new DOMException("message", "name")')`                  |
| `If the implementation has a stack property on normal errors, it also does on DOMExceptions` | PASS    |                                                                                                     |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/248010587)

# Chrome (unstable) #
Testing web-platform-tests at revision 213634be38215a58bbe2ac6bb98db9929d6d2eb5

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/248010588)

# Sauce (MicrosoftEdge) #

**Comment by sideshowbarker** [comment]
w3c-test:mirror

**Comment by rakuco** [comment]
> Chrome fails the stack property test (which is a "should" in the proposed spec) and the toString() inheritance test

The toString() part is easy to fix, but the stack property test needs some work in V8, which currently registers it in %Error% instead of %ErrorPrototype% (I'm assuming we do want to move towards the latter given https://tc39.github.io/proposal-error-stacks/)

**Comment by rakuco** [comment]
Don't you also need to update some of the other tests?
* `constructor-object.js` is checking if DOMException inherits from %Error% and that `name` and `code` are not set in DOMException.prototype.
* `exceptions.html` is checking for the wrong property descriptors and in the wrong location.

**Comment by domenic** [comment]
Great catches, thanks. Let me update. Also I should probably make these tests .any.js tests.

**Comment by foolip** [comment]
@bobholt, is the "No tests run" for 3/4 browsers in this PR a known issue? If not, can you file one with the infra label?

**Comment by domenic** [comment]
@rakuco since the spec PR is merged, this is ready for review, when you have time :)

**Comment by rakuco** [comment]
Sorry, I didn't know you were waiting for me. The change looks fine on my side, and Chromium passes all the new tests (modulo those you'd mentioned before). One of the 3 reviewers need to lgtm the PR though.

**Comment by bobholt** [comment]
@foolip [This PR](https://github.com/w3c/web-platform-tests/pull/6450) addressed an issue that caused only Firefox to return results. That's my guess about what went on here. I've restarted the build to see if that fixed this particular issue.

**Comment by rakuco** [comment]
@tobie @yuki3 @jensl can any of you take a look and approve the PR?

File: WebIDL/ecmascript-binding/es-exceptions/DOMException-constructor-behavior.any.js [hunk 1]
```diff
@@ -135,5 +127,3 @@
   assert_equals(ex.code, 0,
```
Comment (yuki3) [inline WebIDL/ecmascript-binding/es-exceptions/DOMException-constructor-behavior.any.js]: Is this test case ("bar", "UnknownError") a dupe of line 81 ("bar", "foo")?

File: WebIDL/ecmascript-binding/es-exceptions/DOMException-custom-bindings.any.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  assert_equals(Object.getPrototypeOf(DOMException), Function.prototype);
+}, "does not inherit from Error: class-side");
+
+test(() => {
+  const e = new DOMException("message", "name");
+  assert_false(e.hasOwnProperty("message"), "property is not own");
+
+  const propDesc = Object.getOwnPropertyDescriptor(DOMException.prototype, "message");
+  assert_equals(typeof propDesc.get, "function", "property descriptor is a getter");
+  assert_equals(propDesc.set, undefined, "property descriptor is not a getter");
```
Comment (yuki3) [inline WebIDL/ecmascript-binding/es-exceptions/DOMException-custom-bindings.any.js]: Did you mean "not a setter"?
Ditto for all belows.

**Review by rakuco — COMMENTED** [review]
Er, it turns out I can actually approve the change, sorry for the brain fart :-) Yuki's question makes sense though, so not doing it just yet.

**Review by yuki3 — APPROVED** [review]
LGTM with nits.

**Review by rakuco — APPROVED** [review]

**Review by jgraham — APPROVED** [review]

**Pull request merged**